## 01. Importing libraries & ***ords_prods_merge*** dataframe

In [74]:
# Import libraries

import pandas as pd
import numpy as np
import os

#creating path
path=r'C:\Users\Olufemi\Downloads\CareerFoundry Data Analyics\Instacart Basket Analysis 03_02_2025'

#orders_products_merge dataframe
ords_prods_merge=pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_updated_A.pkl'))

# TASK 2

## 02. Creating a Subset

In [92]:
# creating a subset
ords_prods_merge_sub=ords_prods_merge[:1000000]

### 03. Subset Aggregate Mean using groupby() and agg() functions

In [84]:
# Aggregate mean of subset dataframe
ords_prods_merge_sub.groupby(['department_id']).agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                14.800024
2                17.091743
3                17.913544
4                17.893092
5                15.214270
6                15.382135
7                17.694027
8                16.458105
9                15.957363
10               20.091818
11               16.482026
12               15.615061
13               16.484023
14               17.524632
15               15.691875
16               18.014071
17               16.150593
18               19.602850
19               17.631340
20               17.138607
21               21.956893

## 04. Entire ***ords_prods_merge*** Aggregate Mean using groupby() and agg() functions

In [78]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# TASK 3

## Analyzing the Result of Task 2

### The Differences in Mean Order Numbers

Looking at the two results, we can see that the average order number for most departments is higher in the subset compared to the full dataset. Here’s what the result reveals:

1. **The subset contains higher order numbers on average:** this suggests that the subset might be made up of a specific group of customers—perhaps more frequent shoppers or people who tend to place more orders.

2. **The ranking of departments is mostly similar, but the numbers are different.**  
 
3. **Why do the numbers differ?**  
   - The subset might include only certain customers or types of purchases, which skews the numbers higher.  
   - It could be filtered to include only recent orders or exclude certain types of products.  
  
### Key Takeaway
The full dataset gives a general picture of how often different departments are ordered from, while the subset focuses on a specific group that tends to have higher order numbers.


# TASK 4

## 05. Creating a loyalty flag for existing customers

In [101]:
# Creating new column "max_order"

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\Olufemi\AppData\Local\Temp\ipykernel_79072\1980913435.py:3: FutureWarning: The provided callable <function max at 0x000001FDB8E58F40> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [103]:
# Checking for new max_order column created

ords_prods_merge.head(20)

Unnamed: 0  order_id  user_id  order_number  order_dow  order_hour_of_day  \
0            0   2539329        1             1          2                  8   
1            0   2539329        1             1          2                  8   
2            0   2539329        1             1          2                  8   
3            0   2539329        1             1          2                  8   
4            0   2539329        1             1          2                  8   
5            1   2398795        1             2          3                  7   
6            1   2398795        1             2          3                  7   
7            1   2398795        1             2          3                  7   
8            1   2398795        1             2          3                  7   
9            1   2398795        1             2          3                  7   
10           1   2398795        1             2          3                  7   
11           2    473747        1             3          3                 12   
12           2    473747        1             3          3                 12   
13           2    473747        1             3          3                 12   
14           2    473747        1             3          3                 12   
15           2    473747        1             3          3                 12   
16           3   2254736        1             4          4                  7   
17           3   2254736        1             4          4                  7   
18           3   2254736        1             4          4                  7   
19           3   2254736        1             4          4                  7   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                        0         196                  1          0   
1                        0       14084                  2          0   
2                        0       12427                  3          0   
3                        0       26088                  4          0   
4                        0       26405                  5          0   
5                       15         196                  1          1   
6                       15       10258                  2          0   
7                       15       12427                  3          1   
8                       15       13176                  4          0   
9                       15       26088                  5          1   
10                      15       13032                  6          0   
11                      21         196                  1          1   
12                      21       12427                  2          1   
13                      21       10258                  3          1   
14                      21       25133                  4          0   
15                      21       30450                  5          0   
16                      29         196                  1          1   
17                      29       12427                  2          1   
18                      29       10258                  3          1   
19                      29       25133                  4          1   

                               product_name  aisle_id  department_id  prices  \
0                                      Soda        77              7     9.0   
1   Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                       Original Beef Jerky        23             19     4.4   
3                Aged White Cheddar Popcorn        23             19     4.7   
4          XL Pick-A-Size Paper Towel Rolls        54             17     1.0   
5                                      Soda        77              7     9.0   
6                                Pistachios       117             19     3.0   
7                       Original Beef Jerky        23             19     4.4   
8                    Bag of Organic Bananas        24              4 

# 06. Assigning loyalty flag to the customers

In [126]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [128]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [130]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

# 07. Checking the loyalty flags

In [132]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [136]:
# Checking the Loyalty flag in dataframe
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number','max_order']].head(20)

user_id  loyalty_flag  order_number  max_order
0         1  New customer             1         10
1         1  New customer             1         10
2         1  New customer             1         10
3         1  New customer             1         10
4         1  New customer             1         10
5         1  New customer             2         10
6         1  New customer             2         10
7         1  New customer             2         10
8         1  New customer             2         10
9         1  New customer             2         10
10        1  New customer             2         10
11        1  New customer             3         10
12        1  New customer             3         10
13        1  New customer             3         10
14        1  New customer             3         10
15        1  New customer             3         10
16        1  New customer             4         10
17        1  New customer             4         10
18        1  New customer             4         10
19        1  New customer             4         10

# TASK 5

## 08. Determine if the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [139]:
# Finding the min, max and mean of each loyalty flag
ords_prods_merge.groupby(['loyalty_flag']).agg({'prices':['mean','min','max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773487  1.0  25.0
New customer      7.801077  1.0  25.0
Regular customer  7.798145  1.0  25.0

## Insight

### Average Prices:
- New customers have the highest average price (7.801077).
- Regular customers have a slightly lower average price (7.798145).
- Loyal customers have the lowest average price (7.773487)

# TASK 6

## Determining the lower and higher spenders

In [147]:
# Creating column for average spender
ords_prods_merge['average_spend']=ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\Olufemi\AppData\Local\Temp\ipykernel_79072\655122787.py:2: FutureWarning: The provided callable <function mean at 0x000001FDB8E59940> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['average_spend']=ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [149]:
# Creating flag for low and high spender
ords_prods_merge.loc[ords_prods_merge['average_spend']<10,'spender_flag']='Low_spender'

ords_prods_merge.loc[ords_prods_merge['average_spend']>= 10, 'spender_flag'] = 'High_spender'

In [151]:
# Check spender flags
ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
Low_spender     32285150
High_spender      119709
Name: count, dtype: int64

In [153]:
ords_prods_merge[['user_id', 'average_spend', 'spender_flag', 'prices']].head(60)

user_id  average_spend spender_flag  prices
0         1       6.367797  Low_spender     9.0
1         1       6.367797  Low_spender    12.5
2         1       6.367797  Low_spender     4.4
3         1       6.367797  Low_spender     4.7
4         1       6.367797  Low_spender     1.0
5         1       6.367797  Low_spender     9.0
6         1       6.367797  Low_spender     3.0
7         1       6.367797  Low_spender     4.4
8         1       6.367797  Low_spender    10.3
9         1       6.367797  Low_spender     4.7
10        1       6.367797  Low_spender     4.0
11        1       6.367797  Low_spender     9.0
12        1       6.367797  Low_spender     4.4
13        1       6.367797  Low_spender     3.0
14        1       6.367797  Low_spender     8.6
15        1       6.367797  Low_spender     3.2
16        1       6.367797  Low_spender     9.0
17        1       6.367797  Low_spender     4.4
18        1       6.367797  Low_spender     3.0
19        1       6.367797  Low_spender     8.6
20        1       6.367797  Low_spender     1.0
21        1       6.367797  Low_spender     9.0
22        1       6.367797  Low_spender     4.4
23        1       6.367797  Low_spender     3.0
24        1       6.367797  Low_spender     8.6
25        1       6.367797  Low_spender     2.7
26        1       6.367797  Low_spender    14.0
27        1       6.367797  Low_spender     5.0
28        1       6.367797  Low_spender    10.3
29        1       6.367797  Low_spender     9.0
30        1       6.367797  Low_spender     4.4
31        1       6.367797  Low_spender     3.0
32        1       6.367797  Low_spender     8.6
33        1       6.367797  Low_spender     9.0
34        1       6.367797  Low_spender     3.0
35        1       6.367797  Low_spender     4.4
36        1       6.367797  Low_spender     8.6
37        1       6.367797  Low_spender     4.0
38        1       6.367797  Low_spender     4.4
39        1       6.367797  Low_spender     9.0
40        1       6.367797  Low_spender     3.0
41        1       6.367797  Low_spender     8.6
42        1       6.367797  Low_spender    13.4
43        1       6.367797  Low_spender     1.8
44        1       6.367797  Low_spender     1.8
45        1       6.367797  Low_spender    13.4
46        1       6.367797  Low_spender     8.6
47        1       6.367797  Low_spender     9.0
48        1       6.367797  Low_spender     3.0
49        1       6.367797  Low_spender     4.4
50        1       6.367797  Low_spender     9.0
51        1       6.367797  Low_spender    13.4
52        1       6.367797  Low_spender     6.8
53        1       6.367797  Low_spender    12.6
54        1       6.367797  Low_spender     8.6
55        1       6.367797  Low_spender     3.0
56        1       6.367797  Low_spender     1.3
57        1       6.367797  Low_spender     4.0
58        1       6.367797  Low_spender     4.4
59        2       7.515897  Low_spender     5.2

# TASK 7

## Creating an order frequency flag

In [156]:
# Create a column for frequet users
ords_prods_merge['Customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\Olufemi\AppData\Local\Temp\ipykernel_79072\3165240341.py:2: FutureWarning: The provided callable <function median at 0x000001FDB9018680> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['Customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [158]:
# Creating flag on order frequency
ords_prods_merge.loc[ords_prods_merge['Customer_frequency'] > 20, 'frequency_flag'] = 'Non_frequent_customer'

ords_prods_merge.loc[(ords_prods_merge['Customer_frequency']>10) & (ords_prods_merge['Customer_frequency']<=20),'frequency_flag'] = 'Regular customer'

ords_prods_merge.loc[ords_prods_merge['Customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent Customer'

In [160]:
# Check frequency flags

ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent Customer        22796659
Regular customer          6921472
Non_frequent_customer     2686728
Name: count, dtype: int64

In [162]:
ords_prods_merge[['user_id','days_since_prior_order', 'Customer_frequency', 'frequency_flag']].head(60)

user_id  days_since_prior_order  Customer_frequency    frequency_flag
0         1                       0                20.0  Regular customer
1         1                       0                20.0  Regular customer
2         1                       0                20.0  Regular customer
3         1                       0                20.0  Regular customer
4         1                       0                20.0  Regular customer
5         1                      15                20.0  Regular customer
6         1                      15                20.0  Regular customer
7         1                      15                20.0  Regular customer
8         1                      15                20.0  Regular customer
9         1                      15                20.0  Regular customer
10        1                      15                20.0  Regular customer
11        1                      21                20.0  Regular customer
12        1                      21                20.0  Regular customer
13        1                      21                20.0  Regular customer
14        1                      21                20.0  Regular customer
15        1                      21                20.0  Regular customer
16        1                      29                20.0  Regular customer
17        1                      29                20.0  Regular customer
18        1                      29                20.0  Regular customer
19        1                      29                20.0  Regular customer
20        1                      29                20.0  Regular customer
21        1                      28                20.0  Regular customer
22        1                      28                20.0  Regular customer
23        1                      28                20.0  Regular customer
24        1                      28                20.0  Regular customer
25        1                      28                20.0  Regular customer
26        1                      28                20.0  Regular customer
27        1                      28                20.0  Regular customer
28        1                      28                20.0  Regular customer
29        1                      19                20.0  Regular customer
30        1                      19                20.0  Regular customer
31        1                      19                20.0  Regular customer
32        1                      19                20.0  Regular customer
33        1                      20                20.0  Regular customer
34        1                      20                20.0  Regular customer
35        1                      20                20.0  Regular customer
36        1                      20                20.0  Regular customer
37        1                      20                20.0  Regular customer
38        1                      14                20.0  Regular customer
39        1                      14                20.0  Regular customer
40        1                      14                20.0  Regular customer
41        1                      14                20.0  Regular customer
42        1                      14                20.0  Regular customer
43        1                      14                20.0  Regular customer
44        1                       0                20.0  Regular customer
45        1                       0                20.0  Regular customer
46        1                       0                20.0  Regular customer
47        1                       0                20.0  Regular customer
48        1                       0                20.0  Regular customer
49        1                       0                20.0  Regular customer
50        1                      30                20.0  Regular customer
51        1                      30                20.0  Regular customer
52        1                      30                20.0  Regular customer
53      

# TASK 9 EXPORTING DATAFRAME



In [165]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'aggregated_ords_prods_merge.pkl'))